flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/ElasticsearchSinkBase.java

```diff
@@ -165,20 +286,36 @@ public void beforeBulk(long executionId, BulkRequest request) { }
 				@Override
 				public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
 					if (response.hasFailures()) {
-						for (BulkItemResponse itemResp : response.getItems()) {
-							Throwable failure = callBridge.extractFailureCauseFromBulkItemResponse(itemResp);
+						BulkItemResponse itemResponse;
+						Throwable failure;
+
+						for (int i = 0; i < response.getItems().length; i++) {
+							itemResponse = response.getItems()[i];
+							failure = callBridge.extractFailureCauseFromBulkItemResponse(itemResponse);
 							if (failure != null) {
-								LOG.error("Failed Elasticsearch item request: {}", failure.getMessage(), failure);
-								failureThrowable.compareAndSet(null, failure);
+								LOG.error("Failed Elasticsearch item request: {}", itemResponse.getFailureMessage(), failure);
+
+								if (failureHandler.onFailure(request.requests().get(i), failure, requestIndexer)) {
```

Comment:
For the use case you mentioned, that would mean the user implements a stateful `ActionRequestFailureHandler`, with its state being the number of failures so far, correct?

I didn't think about this too much, but I guess there shouldn't be a problem for this.